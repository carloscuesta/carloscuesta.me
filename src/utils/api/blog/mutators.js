// @flow
import { format, formatDistanceToNow } from 'date-fns'
import readingTime from 'reading-time'

export type Post = {
  dateModified: string,
  datePublished: { formatInWords: string, formatDate: string, value: string },
  disqusIdentifier: string,
  excerpt: string,
  html: string,
  images: { featured: { src: string }, preview: { lqpi: string, src: string } },
  readingTime: string,
  slug: string,
  title: string
}

export type PostPreview = $Diff<
  Post,
  {
    dateModified: string,
    disqusIdentifier: string,
    html: string,
    readingTime: string
  }
>

type Payload = { data: Object, html: Object, slug: string }

export const transformPost = (payload: Payload): Post => ({
  dateModified: payload.data.dateModified,
  datePublished: {
    formatInWords: formatDistanceToNow(
      new Date(payload.data.datePublished + ' GMT+2'),
      { addSuffix: true }
    ),
    formatDate: format(
      new Date(payload.data.datePublished),
      'dd MMMM y'
    ),
    value: payload.data.datePublished
  },
  disqusIdentifier: payload.data.disqusIdentifier,
  excerpt: payload.data.excerpt,
  html: payload.html.toString(),
  images: {
    featured: {
      src: payload.data.image
    },
    preview: {
      src: payload.data.image.replace('/upload/', '/upload/w_500/'),
      lqpi: payload.data.image.replace('/upload/', '/upload/t_post-preview-lqpi/')
    }
  },
  readingTime: readingTime(payload.html.toString()).text,
  slug: payload.slug,
  title: payload.data.title
})
