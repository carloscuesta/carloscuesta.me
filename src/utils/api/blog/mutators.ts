import { format, formatDistanceToNow } from 'date-fns'
import { getPlaiceholder } from 'plaiceholder'
import readingTime from 'reading-time'
import type { VFile } from 'vfile'

export type Post = {
  dateModified: string
  datePublished: {
    formatInWords: string
    formatDate: string
    formatMonthDay: string
    value: string
  }
  disqusIdentifier: string
  excerpt: string
  html: string
  images: { featured: { src: string }; preview: { lqpi: string; src: string } }
  readingTime: string
  slug: string
  title: string
}

export type PostPreview = Omit<
  Post,
  'dateModified' | 'disqusIdentifier' | 'html' | 'readingTime'
>

type Payload = {
  data: { [key: string]: string }
  html: VFile
  slug: string
}

export const transformPost = async (payload: Payload): Promise<Post> => {
  const { base64 } = await getPlaiceholder(
    payload.data.image.replace(
      '/upload/',
      '/upload/t_post-preview-lqpi-background-color/',
    ),
  )

  return {
    dateModified: payload.data.dateModified,
    datePublished: {
      formatInWords: formatDistanceToNow(
        new Date(payload.data.datePublished + ' GMT+2'),
        { addSuffix: true },
      ),
      formatDate: format(new Date(payload.data.datePublished), 'dd MMMM y'),
      formatMonthDay: format(new Date(payload.data.datePublished), 'MMM dd'),
      value: payload.data.datePublished,
    },
    disqusIdentifier: payload.data.disqusIdentifier,
    excerpt: payload.data.excerpt,
    html: payload.html.toString(),
    images: {
      featured: {
        src: payload.data.image,
      },
      preview: {
        src: payload.data.image.replace('/upload/', '/upload/w_500/'),
        lqpi: base64,
      },
    },
    readingTime: readingTime(payload.html.toString()).text,
    slug: payload.slug,
    title: payload.data.title,
  }
}
