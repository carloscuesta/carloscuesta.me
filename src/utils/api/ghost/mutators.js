// @flow
import stripTags from 'striptags'
import moment from 'moment'

type Post = {
  excerpt: string,
  image: string,
  publishedAt: { format: string, value: string },
  slug: string,
  tags: Array<string>,
  title: string
}

export const transformPosts = (payload: Object): Array<Post> => payload.posts.map((post) => ({
  excerpt: `${stripTags(post.html).substring(0, 120)}...`,
  image: post.feature_image.replace('/upload/', '/upload/w_500/'),
  publishedAt: {
    format: moment(post.published_at).startOf('hour').fromNow(),
    value: post.published_at
  },
  slug: `/blog/${post.slug}`,
  tags: post.tags.map((tag) => tag.name),
  title: post.title
}))
