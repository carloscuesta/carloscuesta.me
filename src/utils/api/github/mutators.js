// @flow
export type Repository = {
  description: string,
  language?: string,
  name: string,
  stars: number,
  url: string
}

export const transformRepositories = (payload: Array<Object>): Array<Repository> => {
  return payload.map((repo) => ({
    description: repo.description,
    language: repo.language,
    name: repo.name,
    stars: repo.stargazers_count,
    url: repo.html_url
  }))
}

export type UserInformation = {
  followers: number,
  repositories: number
}

export const transformUserInformation = (payload: Object): UserInformation => ({
  followers: payload.followers,
  repositories: payload.public_repos
})
