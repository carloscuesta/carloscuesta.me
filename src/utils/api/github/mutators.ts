export type Repository = {
  description: string,
  language?: string,
  name: string,
  stars: number,
  url: string
}

export const transformRepositories = (
  payload: {
    description: string,
    language: string,
    name: string,
    stargazers_count: number,
    html_url: string
  }[]
): Repository[] => {
  return payload.map((repo) => ({
    description: repo.description,
    language: repo.language,
    name: repo.name,
    stars: repo.stargazers_count,
    url: repo.html_url
  }))
}

export type UserInformation = {
  followers: number
}

export const transformUserInformation = (
  payload: UserInformation
): UserInformation => ({
  followers: payload.followers
})
