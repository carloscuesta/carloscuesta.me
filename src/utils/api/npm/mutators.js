// @flow
export const transformPublishedPackages = (payload: { objects: Array<Object> }): Array<string> => {
  return payload.objects.map((pkg) => pkg.package.name)
}

export const transformDownloadsCount = (payload: Object): number => {
  const packages: Array<Object> = Object.values(payload)

  return packages
    .map((pkg) => pkg.downloads)
    .reduce((memo, value) => memo + value, 0)
}
