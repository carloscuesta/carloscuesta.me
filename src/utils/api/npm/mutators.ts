export const transformPublishedPackages = (
  payload: { objects: Array<{ package: { name: string } }>
}) => {
  return payload.objects.map((pkg) => pkg.package.name)
}

export const transformDownloadsCount = (
  payload: { [key: string]: { downloads: number } }
) => {
  const packages = Object.values(payload)

  return packages
    .map((pkg) => pkg.downloads)
    .reduce((memo, value) => memo + value, 0)
}
