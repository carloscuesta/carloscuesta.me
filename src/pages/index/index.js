// @flow
import React from 'react'
import { NextSeo } from 'next-seo'

import { fetchPosts } from 'src/utils/api/blog'
import { fetchRepositories } from 'src/utils/api/github'
import { type Post } from 'src/utils/api/blog/mutators'
import { type Repository } from 'src/utils/api/github/mutators'
import Header from 'src/components/pages/index/Header'
import Biography from 'src/components/pages/index/Biography'
import Writings from 'src/components/pages/index/Writings'
import OpenSource from 'src/components/pages/index/OpenSource'

type Props = {
  posts: Array<Post>,
  repositories: Array<Repository>
}

const Index = (props: Props) => (
  <>
    <NextSeo canonical='https://carloscuesta.me' />
    <Header />
    <main>
      <Biography />
      <Writings posts={props.posts} />
      <OpenSource repositories={props.repositories} />
    </main>
  </>
)

export const getStaticProps = async (): Promise<{ props: Props }> => {
  const [posts, repositories] = await Promise.all([
    fetchPosts(),
    fetchRepositories()
  ])

  return {
    props: {
      posts: posts.slice(0, 6),
      repositories
    }
  }
}

export default Index
