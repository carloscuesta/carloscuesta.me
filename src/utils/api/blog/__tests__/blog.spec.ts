import { getPostSlugs, fetchPost, fetchPosts } from '../index'
import { transformPost } from '../mutators'
import { VFile } from 'vfile'

import postFixture from './fixtures/post.json'

/* Mock Date.now to avoid updating snapshots for formatDistanceToNow */
Date.now = jest.fn(() => 1593025862540)

describe('Ghost API Client', () => {
  describe('getPostSlugs', () => {
    it('should match getPostSlugs', () => {
      expect(getPostSlugs()).toMatchSnapshot()
    })
  })

  describe('fetchPost', () => {
    it('should return a post with the slug terminal-setup', async () => {
      const post = await fetchPost('terminal-setup')

      expect(post).toMatchSnapshot()
    })
  })

  describe('fetchPosts', () => {
    it('should return an array of posts sorted by datePublished', async () => {
      const posts = await fetchPosts()

      expect(posts.slice(0, 3)).toMatchSnapshot()
    })
  })

  describe('mutators', () => {
    describe('transformPost', () => {
      it('should match the mutated post', () => {
        expect(
          transformPost({
            ...postFixture,
            html: new VFile(postFixture.html),
          })
        ).toMatchSnapshot()
      })
    })
  })
})
