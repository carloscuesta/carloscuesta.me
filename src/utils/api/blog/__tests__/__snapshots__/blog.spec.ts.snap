// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ghost API Client fetchPost should return a post with the slug terminal-setup 1`] = `
{
  "dateModified": "2016-02-06 15:55",
  "datePublished": {
    "formatDate": "06 February 2016",
    "formatInWords": "over 4 years ago",
    "value": "2016-02-06 15:55",
  },
  "disqusIdentifier": "7",
  "excerpt": "Carlos Cuesta terminal setup with ZSH and Oh-My-ZSH, using materialshell and my dotfiles.",
  "html": "<p><strong>Terminal</strong> is a place where a developer spends a lot of time, running commands, looking for outputs and doing a bunch of stuff, so better to stay comfortable and productive with it.<h3 id=shell><a href=#shell aria-hidden=true tabindex=-1><span class=headingLink>#</span></a>Shell</h3><p>I use <a href=http://www.zsh.org rel=nofollow><strong>ZSH</strong></a> as my main shell along with <a href=http://ohmyz.sh rel=nofollow><strong>Oh-My-Zsh</strong></a>. By default, OS X comes with bash, a simpler shell. To setup ZSH and install Oh-My-Zsh run:<pre class=language-language-bash><code class=language-language-bash>$ chsh -s /bin/zsh
$ sh -c "$(curl -fsSL https://raw.github.com/robbyrussell/oh-my-zsh/master/tools/install.sh)"
</code></pre><p>To see what shell are you using run:<pre class=language-language-bash><code class=language-language-bash>$ echo $SHELL
</code></pre><h3 id=ui><a href=#ui aria-hidden=true tabindex=-1><span class=headingLink>#</span></a>UI</h3><p>After setting up ZSH and Oh-My-Zsh, now we are ready to customize and put our shell on steroids.<p>I use a theme called <a href=https://github.com/carloscuesta/materialshell rel=nofollow><strong>materialshell</strong></a> that I've made for my own.<p><a href=https://github.com/carloscuesta/materialshell/blob/master/README.md rel=nofollow><strong>Follow this guide to install and setup my terminal</strong>.</a><p><img alt="Carlos Cuesta Shell - MaterialShell"src=https://res.cloudinary.com/carloscuesta/image/upload/v1454753967/gazne7uhxuq8ugpfvile.png><h3 id=dotfiles><a href=#dotfiles aria-hidden=true tabindex=-1><span class=headingLink>#</span></a>Dotfiles</h3><p>I have my own <a href=https://github.com/carloscuesta/dotfiles rel=nofollow><strong>dotfiles repository</strong></a> with all my shell configuration files, where you can take a look to learn and be more productive with your shell.",
  "images": {
    "featured": {
      "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1593531858/blog-featured-images/Terminal_Setup.png",
    },
    "preview": {
      "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_post-preview-lqpi/v1593531858/blog-featured-images/Terminal_Setup.png",
      "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1593531858/blog-featured-images/Terminal_Setup.png",
    },
  },
  "readingTime": "1 min read",
  "slug": "terminal-setup",
  "title": "Terminal setup",
}
`;

exports[`Ghost API Client fetchPosts should return an array of posts sorted by datePublished 1`] = `
[
  {
    "datePublished": {
      "formatDate": "29 January 2023",
      "formatInWords": "in over 2 years",
      "value": "2023-01-29 22:00",
    },
    "excerpt": "A codemod is a tool that helps you with large-scale refactors. In this article I'm going to explain how you can use them along with Abstract Syntax Trees.",
    "images": {
      "featured": {
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1675022732/blog-featured-images/Effective_Refactoring_with_Codemods.png",
      },
      "preview": {
        "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_post-preview-lqpi/v1675022732/blog-featured-images/Effective_Refactoring_with_Codemods.png",
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1675022732/blog-featured-images/Effective_Refactoring_with_Codemods.png",
      },
    },
    "slug": "effective-refactoring-with-codemods",
    "title": "Effective Refactoring with Codemods",
  },
  {
    "datePublished": {
      "formatDate": "23 May 2022",
      "formatInWords": "in almost 2 years",
      "value": "2022-05-23 17:55",
    },
    "excerpt": "Let's walk through the experience and the outcome of implementing a tool to generate code. What are the benefits, the problems they solve and how to implement them.",
    "images": {
      "featured": {
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1653167435/blog-featured-images/Code_generators.png",
      },
      "preview": {
        "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_post-preview-lqpi/v1653167435/blog-featured-images/Code_generators.png",
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1653167435/blog-featured-images/Code_generators.png",
      },
    },
    "slug": "using-generators-to-improve-developer-productivity",
    "title": "Using generators to improve developer productivity",
  },
  {
    "datePublished": {
      "formatDate": "01 March 2022",
      "formatInWords": "in over 1 year",
      "value": "2022-03-01 22:00",
    },
    "excerpt": "An architecture decision record also known as ADR, is a document that captures an important architecture decision, including the context and the consequences.",
    "images": {
      "featured": {
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1646170453/blog-featured-images/writing-adrs.png",
      },
      "preview": {
        "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_post-preview-lqpi/v1646170453/blog-featured-images/writing-adrs.png",
        "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1646170453/blog-featured-images/writing-adrs.png",
      },
    },
    "slug": "writing-architecture-decision-records",
    "title": "Writing Architecture Decision Records",
  },
]
`;

exports[`Ghost API Client getPostSlugs should match getPostSlugs 1`] = `
[
  "2016-year-in-review",
  "chrome-devtools-customization",
  "composing-uis-with-finite-state-machines",
  "debugging-with-the-javascript-console",
  "discovering-hyperterm",
  "effective-refactoring-with-codemods",
  "front-end-tooling-guide-to-improve-developer-experience",
  "functional-programming-in-javascript",
  "git-development-workflow",
  "greenkeeper",
  "hello-world",
  "how-to-deploy-static-sites-with-surge",
  "https-cloudflare",
  "immutability-in-javascript",
  "javascript-monorepos-lerna-yarn-workspaces",
  "keep-npm-dependencies-up-to-date",
  "managing-react-native-crashes-with-error-boundaries",
  "mastering-and-hacking-slack",
  "moving-to-digitalocean",
  "moving-ulabox-to-nextjs",
  "react-amsterdam-2017",
  "react-native-meets-codepush",
  "reactiveconf-2017",
  "scalable-react-components-architecture",
  "shipping-react-native-apps-with-fastlane",
  "shipping-react-native-fastlane-travis",
  "terminal-setup",
  "using-generators-to-improve-developer-productivity",
  "working-remotely-as-a-software-engineer",
  "writing-architecture-decision-records",
]
`;

exports[`Ghost API Client mutators transformPost should match the mutated post 1`] = `
{
  "dateModified": "2016-02-06 15:55",
  "datePublished": {
    "formatDate": "06 February 2016",
    "formatInWords": "over 4 years ago",
    "value": "2016-02-06 15:55",
  },
  "disqusIdentifier": "7",
  "excerpt": "Carlos Cuesta terminal setup with ZSH and Oh-My-ZSH, using materialshell and my dotfiles.",
  "html": "",
  "images": {
    "featured": {
      "src": "https://res.cloudinary.com/carloscuesta/image/upload/v1454754838/kcuq34zbo1yieygmx0y6.png",
    },
    "preview": {
      "lqpi": "https://res.cloudinary.com/carloscuesta/image/upload/t_post-preview-lqpi/v1454754838/kcuq34zbo1yieygmx0y6.png",
      "src": "https://res.cloudinary.com/carloscuesta/image/upload/w_500/v1454754838/kcuq34zbo1yieygmx0y6.png",
    },
  },
  "readingTime": "0 min read",
  "slug": "terminal-setup",
  "title": "Terminal setup",
}
`;
